File: 230-kth-smallest-element-in-a-bst/230-kth-smallest-element-in-a-bst.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    public int counter;
    int solve(TreeNode root){
        
        if(root == null){
            return -1;
        }
        else if(counter == 0){
            return root.val;
        }
        
        int l = solve(root.left);
        if(l != -1){
            return l;
        }
        counter--;
        if(counter == 0){
            return root.val;
        }
        int r = solve(root.right);
        if(r != -1){
            return r;
        }
        return -1;
    }
    public int kthSmallest(TreeNode root, int k) {
        counter = k;
        return solve(root);
    }
}

File: 230-kth-smallest-element-in-a-bst/230-kth-smallest-element-in-a-bst.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    TreeNode* kthSmallestUtil(TreeNode *root,int k,int &countNode) {
        if(root == NULL) {
            return NULL;
        }
        
        TreeNode *l = kthSmallestUtil(root->left,k,countNode);
       
        if(l != NULL) {
            return l;
        }
        
        countNode++;
        if(countNode == k) {
            return root;
        }
        
        TreeNode *r = kthSmallestUtil(root->right,k,countNode);
        
        return r;
    }
    int kthSmallest(TreeNode* root, int k) {
        
        int countNode = 0;
        
        TreeNode *node = kthSmallestUtil(root,k,countNode);
        return node->val;
    }
};